### DISCUSSION TITLE
How to find LineStrings which connect with each other？

### DISCUSSION THREAD

**Discussion by forestbat** [shapely/shapely#1754] [Q&A] [answered]
Such as I have 3 linestrings：
`LINESTRING (0 0, 1 1)`
`LINESTRING (1 1, 2 0)`
`LINESTRING (1 2, 2 3)`
It's obvious that the 1st and 2nd linestring connects with each other, now I want to find `LINESTRING (1 1, 2 0)` by `LINESTRING (0 0, 1,1)` or opposite, so is there any method in shapely can help me complete this mission?

**Comment by brendan-ward** [comment]
What do you mean by "connect"?

Perhaps you mean [intersects](https://shapely.readthedocs.io/en/stable/reference/shapely.intersects.html#shapely.intersects)?

(note: your WKT strings are not valid, you have extra commas)

```Python
>>> import shapely
>>> line1 = shapely.from_wkt('LINESTRING (1 1, 2 0)')
>>> line2 = shapely.from_wkt('LINESTRING (0 0, 1 1)')
>>> shapely.intersects(line1, line2)
True
```

Depending on your goals, you can use numpy broadcasting for one of the inputs:

```Python
>>> line3 = shapely.from_wkt('LINESTRING (1 2, 2 3)')
>>> shapely.intersects(line1, [line2, line3])
array([ True, False])
```

Perhaps you mean [touches](https://shapely.readthedocs.io/en/stable/reference/shapely.touches.html#shapely.touches) if by "connect" you mean they share an endpoint but not any intermediate point or cross.

**Comment by forestbat** [comment]
I want to write a program about geometry analysis，such as an array：`[line6, line4, line1, line2, line3, line5]`，in this array `line1 line2 line3` are connected end to end:

![图片](https://user-images.githubusercontent.com/23413915/217448653-3745a938-a2bb-4e0b-9201-c44ad18266e2.png)

Now I want to start from LINE1 (or LINE1's coords[0]) to find LINE2 and LINE3 until there is no any lines can be find（or start from LINE2 to find LINE1 and LINE3 and further lines， you can think this as deep -first-search），if I use `intersects`，every lines should use `O(n)` time to find other geometries which has intersection with it，so for getting `[line1, line2, line3]` from these linestrings(yes, it has special order), I probably need 18 times of search. If there is too many linestrings (such as Amazon River's basin, you can regard every river as a linestring), performance cost is hard to bear.
And [touches](https://shapely.readthedocs.io/en/stable/reference/shapely.touches.html#shapely.touches) doesn't fit me，because I really want to find linestrings with intermediate point or cross，such as this：

![图片](https://user-images.githubusercontent.com/23413915/217450222-2d661d49-791c-482d-a7d6-35e3eb735598.png)

**Comment by brendan-ward** [comment, marked as answer]
I think you can use a combination of STRtree to find all intersecting pairs and an undirected graph to find connected components, e.g.,  networkx [connected_components](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.components.connected_components.html#networkx.algorithms.components.connected_components)

Something along the lines of (untested / partial implementation):

```Python
>>> import shapely
>>> import numpy as np

# create lines as per above, then put into array
>>> lines = np.array([line1, line2, line3, ...lineN])
>>> tree = shapely.STRtree(lines)

# find all intersections within the tree 
>>> left, right = tree.query(lines, predicate='intersects')

# drop self-intersections
>>> ix = left!=right

# make array of pairs of left index, right index
>>> pairs = np.array([left[ix], right[ix]]).T
```

NOTE: `pairs` has symmetric pairs, e.g., A=>B and B=>A, which is desirable for an undirected graph.  The left and right indexes in pairs are integer indexes into lines.

Once you have intersecting pairs, you build your undirected graph, and retrieve connected components.  Where you have A<=>B and B<=>C, this will give you a component that is A, B, C; this is your connected network of linestrings.

If you need to generate a valid network topology from intersecting lines and preserve information from their original segments (e.g., river or tributary name), that is a bit more involved and depends on your specific needs.  You'll need to cut the lines at their intersection points (various ways to do this), then recalculate the set of connected lines.

Alternatively, if you don't care about the original segments and only want to extract lines from a valid topology, you could use a `union_all` operation, e.g., given indexes for a given component in your graph `component_indexes` (i.e., in a loop over components):

```Python
>>> component_lines = lines.take(component_indexes)
>>> merged = shapely.union_all(component_lines)
```

This should return a valid MultiLineString containing the segments after cutting them at their intersections.
